scripts/test_solidity_lsp.py

```diff
@@ -0,0 +1,624 @@
+        action="store_true",
+        help="Be more verbose by also printing assertions."
+    )
+    parser.add_argument(
+        "-t, --test-pattern",
+        dest="test_pattern",
+        type=str,
+        default="*",
+        help="Filters all available tests by matching against this test pattern (using globbing)",
+        nargs="?"
+    )
+    parser.add_argument(
+        "solc_path",
+        type=str,
+        default="solc",
+        help="Path to solc binary to test against",
+        nargs="?"
+    )
+    parser.add_argument(
+        "project_root_dir",
+        type=str,
+        default=f"{os.path.dirname(os.path.realpath(__file__))}/..",
+        help="Path to Solidity project's root directory (must be fully qualified).",
+        nargs="?"
+    )
+    return parser
+
+class SolidityLSPTestSuite: # {{{
+    tests_passed: int = 0
+    tests_failed: int = 0
+    assertion_count: int = 0   # number of total assertions executed so far
+    assertions_passed: int = 0
+    assertions_failed: int = 0
+    print_assertions: bool = False
+    trace_io: bool = False
+    test_pattern: str
+
+    def __init__(self):
+        colorama.init()
+        args = create_cli_parser().parse_args()
+        self.solc_path = args.solc_path
+        self.project_root_dir = os.path.realpath(args.project_root_dir) + "/test/libsolidity/lsp"
+        self.project_root_uri = "file://" + self.project_root_dir
+        self.print_assertions = args.print_assertions
+        self.trace_io = args.trace_io
+        self.test_pattern = args.test_pattern
+
+        print(f"{SGR_NOTICE}test pattern: {self.test_pattern}{SGR_RESET}")
+
+    def main(self) -> int:
+        """
+        Runs all test cases.
+        Returns 0 on success and the number of failing assertions (capped to 127) otherwise.
+        """
+        all_tests = sorted(
+            [
+                name for name in dir(SolidityLSPTestSuite)
+                if callable(getattr(SolidityLSPTestSuite, name)) and name.startswith("test_")
+            ]
+        )
+        filtered_tests = fnmatch.filter(all_tests, self.test_pattern)
+        for method_name in filtered_tests:
+            test_fn = getattr(self, method_name)
+            title: str = test_fn.__name__[5:]
+            print(f"{SGR_TEST_BEGIN}Testing {title} ...{SGR_RESET}")
+            try:
+                with JsonRpcProcess(self.solc_path, ["--lsp"], trace_io=self.trace_io) as solc:
+                    test_fn(solc)
+                    self.tests_passed = self.tests_passed + 1
+            except ExpectationFailed as e:
+                print(f"{e}")
+                self.tests_failed = self.tests_failed + 1
+
+        print(
+            f"\nSummary:\n\n"
+            f"  Test cases: {self.tests_passed} passed, {self.tests_failed} failed\n"
+            f"  Assertions: {self.assertions_passed} passed, {self.assertions_failed} failed\n"
+        )
+
+        return min(self.assertions_failed, 127)
```

Comment:
Please also return nonzero if we have at least one failed test, regardless of the number of failed assertions.